tfhe/src/integer/server_key/radix_parallel/scalar_sub.rs

```diff
@@ -124,27 +126,469 @@ impl ServerKey {
+
+    fn overflowing_scalar_sub_assign_parallelized_sequential<Scalar>(
+        &self,
+        lhs: &mut RadixCiphertext,
+        scalar: Scalar,
+    ) -> BooleanBlock
+    where
+        Scalar: UnsignedNumeric + DecomposableInto<u8> + std::ops::Not<Output = Scalar>,
+    {
+        assert!(!lhs.blocks.is_empty());
+
+        let scalar_blocks = BlockDecomposer::new(scalar, self.message_modulus().0.ilog2())
+            .iter_as::<u8>()
+            .take(lhs.blocks.len())
+            .collect::<Vec<_>>();
+
+        // If the block does not have a carry after the subtraction, it means it needs to
+        // borrow from the next block
+        let compute_borrow_lut = self.key.generate_lookup_table(|x| {
+            if x < self.message_modulus().0 as u64 {
+                1
+            } else {
+                0
+            }
+        });
+
+        let delta = (1_u64 << 63) / (self.message_modulus().0 * self.carry_modulus().0) as u64;
+        let mut borrow = self.key.create_trivial(0);
+        for (lhs_b, scalar_b) in lhs.blocks.iter_mut().zip(scalar_blocks.iter().copied()) {
+            crate::core_crypto::algorithms::lwe_ciphertext_plaintext_sub_assign(
+                &mut lhs_b.ct,
+                crate::core_crypto::prelude::Plaintext(u64::from(scalar_b) * delta),
+            );
+            crate::core_crypto::algorithms::lwe_ciphertext_plaintext_add_assign(
+                &mut lhs_b.ct,
+                crate::core_crypto::prelude::Plaintext(self.message_modulus().0 as u64 * delta),
+            );
+            crate::core_crypto::algorithms::lwe_ciphertext_sub_assign(&mut lhs_b.ct, &borrow.ct);
+            lhs_b.set_noise_level(lhs_b.noise_level() + borrow.noise_level());
+
+            borrow.clone_from(lhs_b);
+
+            rayon::join(
+                || self.key.message_extract_assign(lhs_b),
+                || {
+                    self.key
+                        .apply_lookup_table_assign(&mut borrow, &compute_borrow_lut)
+                },
+            );
+        }
+
+        if BlockDecomposer::new(scalar, self.message_modulus().0.ilog2())
+            .iter_as::<u8>()
+            .skip(lhs.blocks.len())
+            .any(|scalar_block| scalar_block != 0)
+        {
+            // The value we subtracted is bigger than what the ciphertext
+            // can represent, so its a trivial overflow
+            self.create_trivial_boolean_block(true)
+        } else {
+            // borrow of last block indicates overflow
+            BooleanBlock::new_unchecked(borrow)
+        }
+    }
+
+    fn overflowing_scalar_sub_assign_parallelized_at_least_4_bits<Scalar>(
+        &self,
+        lhs: &mut RadixCiphertext,
+        scalar: Scalar,
+    ) -> BooleanBlock
+    where
+        Scalar: UnsignedNumeric + DecomposableInto<u8> + std::ops::Not<Output = Scalar>,
+    {
+        let packed_modulus = (self.message_modulus().0 * self.message_modulus().0) as u64;
+        assert!(scalar >= Scalar::ZERO);
```

Comment:
hmm ?

Reply:
hehe forgot that there was a UnsignedNumeric bound

Reply:
but still we don't have subtraction of a negative number ?

Reply:
maybe missing the unsigned in the naming of the function

Reply:
Yeah its just missing the unsigned from the naming as we already have the scalar sub (its the algorithm for both signed and unsigned)
for the overflowing sub we have the signed one, and this adds the unsigned one as the detection is not done in the same way